File: step_13_binary_trees/13.2_medium/IB105_invert_binary_tree/solution.py
"""
Problem: Invert Binary Tree (LeetCode #226)
Difficulty: EASY | XP: 10
"""
from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1 & 3: RECURSIVE DFS (Preorder Swap)
# Swap children, then recurse into both subtrees.
# Time: O(n)  |  Space: O(h)
# ============================================================
def invert_recursive(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return None
    root.left, root.right = root.right, root.left
    invert_recursive(root.left)
    invert_recursive(root.right)
    return root


# ============================================================
# APPROACH 2: ITERATIVE BFS
# Level-by-level: dequeue, swap children, enqueue children.
# Time: O(n)  |  Space: O(w)
# ============================================================
def invert_bfs(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return None

    queue = deque([root])
    while queue:
        node = queue.popleft()
        node.left, node.right = node.right, node.left
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
    return root


# ============================================================
# APPROACH 3 (ALT): ITERATIVE DFS with Stack
# Same as BFS but using a stack.
# Time: O(n)  |  Space: O(h)
# ============================================================
def invert_iterative_dfs(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return None

    stack = [root]
    while stack:
        node = stack.pop()
        node.left, node.right = node.right, node.left
        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)
    return root


def build_tree(arr: List[Optional[int]]) -> Optional[TreeNode]:
    if not arr or arr[0] is None:
        return None
    root = TreeNode(arr[0])
    queue = deque([root])
    i = 1
    while queue and i < len(arr):
        curr = queue.popleft()
        if i < len(arr) and arr[i] is not None:
            curr.left = TreeNode(arr[i])
            queue.append(curr.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            curr.right = TreeNode(arr[i])
            queue.append(curr.right)
        i += 1
    return root


def tree_to_list(root: Optional[TreeNode]) -> List[Optional[int]]:
    if root is None:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    # Trim trailing nulls
    while result and result[-1] is None:
        result.pop()
    return result


if __name__ == "__main__":
    print("=== Invert Binary Tree ===\n")

    tests = [
        ([4, 2, 7, 1, 3, 6, 9], [4, 7, 2, 9, 6, 3, 1]),
        ([2, 1, 3], [2, 3, 1]),
        ([1], [1]),
        ([], []),
        ([1, 2], [1, None, 2]),
    ]

    methods = [
        ("Recursive", invert_recursive),
        ("BFS", invert_bfs),
        ("Iterative DFS", invert_iterative_dfs),
    ]

    for arr, expected in tests:
        print(f"Tree: {arr} -> Expected: {expected}")
        for name, fn in methods:
            root = build_tree(arr)
            result = fn(root)
            actual = tree_to_list(result)
            status = "PASS" if actual == expected else "FAIL"
            print(f"  {name}: {actual} {status}")
        print()

File: step_13_binary_trees/13.2_medium/IB105_invert_binary_tree/Solution.java
/**
 * Problem: Invert Binary Tree (LeetCode #226)
 * Difficulty: EASY | XP: 10
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // ============================================================
    // APPROACH 1 & 3: RECURSIVE DFS (Preorder Swap)
    // Swap children, then recurse into both subtrees.
    // Time: O(n)  |  Space: O(h)
    // ============================================================
    static class RecursiveDFS {
        public static TreeNode invertTree(TreeNode root) {
            if (root == null) return null;
            // Swap children
            TreeNode temp = root.left;
            root.left = root.right;
            root.right = temp;
            // Recurse
            invertTree(root.left);
            invertTree(root.right);
            return root;
        }
    }

    // ============================================================
    // APPROACH 2: ITERATIVE BFS
    // Level-by-level: dequeue, swap children, enqueue children.
    // Time: O(n)  |  Space: O(w)
    // ============================================================
    static class IterativeBFS {
        public static TreeNode invertTree(TreeNode root) {
            if (root == null) return null;

            Queue<TreeNode> queue = new LinkedList<>();
            queue.offer(root);

            while (!queue.isEmpty()) {
                TreeNode node = queue.poll();
                // Swap children
                TreeNode temp = node.left;
                node.left = node.right;
                node.right = temp;
                // Enqueue children
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            return root;
        }
    }

    // ============================================================
    // APPROACH 3 (ALT): ITERATIVE DFS with Stack
    // Same as BFS but using a stack.
    // Time: O(n)  |  Space: O(h)
    // ============================================================
    static class IterativeDFS {
        public static TreeNode invertTree(TreeNode root) {
            if (root == null) return null;

            Deque<TreeNode> stack = new ArrayDeque<>();
            stack.push(root);

            while (!stack.isEmpty()) {
                TreeNode node = stack.pop();
                // Swap children
                TreeNode temp = node.left;
                node.left = node.right;
                node.right = temp;
                // Push children
                if (node.left != null) stack.push(node.left);
                if (node.right != null) stack.push(node.right);
            }
            return root;
        }
    }

    // Helper: build tree from level-order array
    public static TreeNode buildTree(Integer[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null) return null;
        TreeNode root = new TreeNode(arr[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;
        while (!queue.isEmpty() && i < arr.length) {
            TreeNode curr = queue.poll();
            if (i < arr.length && arr[i] != null) {
                curr.left = new TreeNode(arr[i]);
                queue.offer(curr.left);
            }
            i++;
            if (i < arr.length && arr[i] != null) {
                curr.right = new TreeNode(arr[i]);
                queue.offer(curr.right);
            }
            i++;
        }
        return root;
    }

    // Helper: convert tree to level-order list for comparison
    public static List<Integer> treeToList(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(node.val);
                queue.offer(node.left);
                queue.offer(node.right);
            } else {
                result.add(null);
            }
        }
        // Trim trailing nulls
        while (!result.isEmpty() && result.get(result.size() - 1) == null) {
            result.remove(result.size() - 1);
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println("=== Invert Binary Tree ===\n");

        Object[][] tests = {
            {new Integer[]{4, 2, 7, 1, 3, 6, 9}, new Integer[]{4, 7, 2, 9, 6, 3, 1}},
            {new Integer[]{2, 1, 3}, new Integer[]{2, 3, 1}},
            {new Integer[]{1}, new Integer[]{1}},
            {new Integer[]{}, new Integer[]{}},
            {new Integer[]{1, 2}, new Integer[]{1, null, 2}},
        };

        String[] methods = {"Recursive", "BFS", "Iterative DFS"};

        for (Object[] test : tests) {
            Integer[] arr = (Integer[]) test[0];
            Integer[] expectedArr = (Integer[]) test[1];
            List<Integer> expected = Arrays.asList(expectedArr);

            // Test each approach on a fresh tree
            for (int m = 0; m < 3; m++) {
                TreeNode root = buildTree(arr);
                TreeNode result;
                switch (m) {
                    case 0: result = RecursiveDFS.invertTree(root); break;
                    case 1: result = IterativeBFS.invertTree(root); break;
                    default: result = IterativeDFS.invertTree(root); break;
                }
                List<Integer> actual = treeToList(result);
                String status = actual.equals(expected) ? "PASS" : "FAIL";
                if (m == 0) {
                    System.out.printf("Tree: %s -> Expected: %s%n", Arrays.toString(arr), expected);
                }
                System.out.printf("  %s: %s %s%n", methods[m], actual, status);
            }
            System.out.println();
        }
    }
}
